### PR TITLE
More Bignum fiddling.

### PR DESCRIPTION
**Pull request by ColinH**
A couple more changes, most noteworthy are the reduction of `used_bigits_` and `exponent_` to 16 bits, together with a wrapper for `exponent_` that ensures that it doesn't overflow (only when assertions are enabled), and I've removed the class invariant that all unused "bigits" are zero so that the constructor, `Zero()` etc. don't need to clear everything all the time.

### CODE REVIEW

**Comment by floitsch** [comment]
Will have to take more time to review this PR. (Would like to make sure the exponent can't exceed a certain range).
Please ping me from time to time (at least once a week), so I don't forget.

**Comment by ColinH** [comment]
Ok will do :-)

(As mentioned in the other thread, during a run of the test-suite the exponent doesn't seem to exceed 50, but of course that's no proof, and it would be more reassuring if you found one.)

**Comment by ColinH** [comment]
Btw, is there a particular reason for using exactly 28 of the Bignum chunks/bigit bits instead of 31? If I'm understanding correctly you're not using 32 in order to have an additional free carry bit.

**Comment by ColinH** [comment]
To answer my own question:

- 29 bits per chunk/bigit seems to work, i.e. the test-suite runs without errors.
- 30 bits triggers an assert in `Bignum::DivideModuloIntBignum()`.
- 31 bits triggers an `UNREACHABLE` in `Bignum::Square()`.

**Comment by ColinH** [comment]
This PR got a bit larger than anticipated, I'll try to explain the commits, in order to, hopefully, make it easier to sort through it all...

*    Update documentation -- README.md still refers to `double-conversion.h` being "heavily documented", but, since this was split, it should now refer to `double-to-string.h` and `string-to-double.h` instead.
*     Generalise conversion -- replaces `Bignum::AssignHexString()` with a shorter implementation that works with non-factor-of-four `kBigitSize`.
*     Allow inlining -- moves three very small member functions of class `Bignum` into the header.
*     Remove superfluous -- removes two redundant `inline` specifiers.
*     Not a class -- converts `PowersOfTenCache` from a "class" into a namespace.
*     For Mac and Emacs -- adds `*~` and `*.dylib` to `.gitignore`.
*     Fix exponent type -- changes the exponent in `diy-fp.h` from `int` to `int32_t`.
*     Undo mistake -- I added an `Sconscript~` by accident, this removes it again.
*     Fix comments -- improves some comments in `diy-fp.h`.
*     Remove Bignum clearing -- changes `Bignum` to not need to clear unused bigits (luckily not much was missing).
*     Assert no Bignum exponent overflow -- wrap `Bignum::exponent_` in a class that asserts that no overflow happens, can possibly be removed again in the future?
*     Consistent naming -- consistently use "bigits".
*     Formatting -- consistenly use braces after `if`.
*     Another const -- add a missing `const`.
*     Remove superfluous -- remove two superfluous calls to `Bignum::EnsureCapacity()`.
*     Compact Bignum sizes -- mostly changes `Bignum::exponent_` and `Bignum::used_bigits_` to `int16_t`.
*     More const -- add a couple of missing/possible `const`.
*     Allow inlining -- move the single function from `diy-fp.cc` to `diy-fp.h`.
*     More const -- add a couple of missing/possible `const`.
*     Char has size 1 -- remove a constant with value `sizeof(char)` since that is guaranteed to be `1`.

**Comment by floitsch** [comment]
Fyi: I started reviewing this PR.
I checked the exponent, and it can't exceed the 16 bits (afaics), so that's good.
Still going through the rest.

**Comment by ColinH** [comment]
If the exponent is "safe" in 16 bits, can I remove `ExponentWrapper` again?

File: double-conversion/bignum.cc [hunk 1]
```diff
@@ -33,148 +33,126 @@
...
-
-  int needed_bigits = length * 4 / kBigitSize + 1;
-  EnsureCapacity(needed_bigits);
-  int string_index = length - 1;
-  for (int i = 0; i < needed_bigits - 1; ++i) {
-    // These bigits are guaranteed to be "full".
-    Chunk current_bigit = 0;
-    for (int j = 0; j < kBigitSize / 4; j++) {
-      current_bigit += HexCharValue(value[string_index--]) << (j * 4);
+  EnsureCapacity( ( ( value.length() * 4 ) + kBigitSize - 1 ) / kBigitSize );  // Could be less if we ignored leading zeros while converting.
```
[the remark this answers is not in the document]
Comment (ColinH) [reply double-conversion/bignum.cc]: Your code formatting style is very different from ours, and though I tried to stick to yours I must have reverted to my usual style once or twice without noticing, will fix now.

File: double-conversion/bignum.cc [hunk 2]
```diff
@@ -33,148 +33,126 @@
...
+  DOUBLE_CONVERSION_ASSERT( sizeof( uint64_t ) * 8 >= kBigitSize + 4 );  // TODO: static_assert
+  uint64_t tmp = 0;  // Accumulate converted hex digits until at least kBigitSize bits, works with non-factor-of-four kBigitSizes.
+  for( int cnt = 0; !value.is_empty(); value.pop_back() ) {
```
[the remark this answers is not in the document]
Comment (ColinH) [reply double-conversion/bignum.cc]: Will be fixed in next commit.

File: double-conversion/bignum.cc [hunk 3]
```diff
@@ -201,24 +179,28 @@ void Bignum::AddBignum(const Bignum& other) {
...
-    Chunk sum = RawBigit(bigit_pos) + carry;
+    const Chunk my = (bigit_pos < used_bigits_) ? RawBigit(bigit_pos) : 0;
+    const Chunk sum = my + carry;
     RawBigit(bigit_pos) = sum & kBigitMask;
     carry = sum >> kBigitSize;
-    bigit_pos++;
+    ++bigit_pos;
   }
-  used_digits_ = (std::max)(bigit_pos, used_digits_);
+  used_bigits_ = (std::max)(bigit_pos, int(used_bigits_));
```
[the remark this answers is not in the document]
Comment (ColinH) [reply double-conversion/bignum.cc]: The automatic variable `bigit_pos` is an `int`, while `used_bigits_` is now an `int16_t`, and `std::max()` doesn't like differing types. I prefer to use constructor-notation `int(value)` over `static_cast<int>(value)` for "safe" conversions, i.e. I leave `static_cast` for the (usually very rare) "something strange going on here" cases that I want to be able to `grep(1)` for. I'll change it to a `static_cast`.

**Comment by floitsch** [comment]
> If the exponent is "safe" in 16 bits, can I remove `ExponentWrapper` again?

Yes. Works for me.

File: double-conversion/bignum.cc [hunk 4]
```diff
@@ -33,148 +33,128 @@
...
-  int string_index = length - 1;
-  for (int i = 0; i < needed_bigits - 1; ++i) {
-    // These bigits are guaranteed to be "full".
-    Chunk current_bigit = 0;
-    for (int j = 0; j < kBigitSize / 4; j++) {
-      current_bigit += HexCharValue(value[string_index--]) << (j * 4);
+  // Required capacity could be reduced by ignoring leading zeros.
+  EnsureCapacity(((value.length() * 4) + kBigitSize - 1) / kBigitSize);
+  DOUBLE_CONVERSION_ASSERT(sizeof(uint64_t) * 8 >= kBigitSize + 4);  // TODO: static_assert
+  uint64_t tmp = 0;  // Accumulates converted hex digits until at least
```
Comment (floitsch) [inline double-conversion/bignum.cc]:
```suggestion
  // Accumulates converted hex digits until at least kBigitSize bits.
  // Works with non-factor-of-four kBigitSizes.
  uint64_t tmp = 0;  // Accumulates converted hex digits until at least
```

File: double-conversion/bignum.cc [hunk 5]
```diff
@@ -33,148 +33,128 @@
...
+  // kBigitSize bits, works with non-factor-of-four kBigitSizes.
```
Comment (floitsch) [inline double-conversion/bignum.cc]:
```suggestion
```

File: double-conversion/bignum.cc [hunk 6]
```diff
@@ -33,148 +33,128 @@
...
-
-  Chunk most_significant_bigit = 0;  // Could be = 0;
-  for (int j = 0; j <= string_index; ++j) {
-    most_significant_bigit <<= 4;
-    most_significant_bigit += HexCharValue(value[j]);
   }
-  if (most_significant_bigit != 0) {
-    RawBigit(used_digits_) = most_significant_bigit;
-    used_digits_++;
+  if(tmp > 0) {
```
Comment (floitsch) [inline double-conversion/bignum.cc]:
```suggestion
  if (tmp > 0) {
```

File: double-conversion/bignum.cc [hunk 7]
```diff
@@ -33,148 +33,128 @@
...
 }
 
 // Guaranteed to lie in one Bigit.
-void Bignum::AssignUInt16(uint16_t value) {
+void Bignum::AssignUInt16(const uint16_t value) {
   DOUBLE_CONVERSION_ASSERT(kBigitSize >= BitSize(value));
   Zero();
-  if (value == 0) {
-    return;
+  if(value > 0 ) {
```
Comment (floitsch) [inline double-conversion/bignum.cc]:
```suggestion
  if (value > 0) {
```

File: double-conversion/bignum.cc [hunk 8]
```diff
@@ -33,148 +33,128 @@
...
-    for (int j = 0; j < kBigitSize / 4; j++) {
-      current_bigit += HexCharValue(value[string_index--]) << (j * 4);
+  // Required capacity could be reduced by ignoring leading zeros.
+  EnsureCapacity(((value.length() * 4) + kBigitSize - 1) / kBigitSize);
+  DOUBLE_CONVERSION_ASSERT(sizeof(uint64_t) * 8 >= kBigitSize + 4);  // TODO: static_assert
+  uint64_t tmp = 0;  // Accumulates converted hex digits until at least
+  // kBigitSize bits, works with non-factor-of-four kBigitSizes.
+  for (int cnt = 0; !value.is_empty(); value.pop_back()) {
+    tmp |= (HexCharValue(value.last()) << cnt);
+    if ((cnt += 4 ) >= kBigitSize) {
```
Comment (floitsch) [inline double-conversion/bignum.cc]:
```suggestion
    if ((cnt += 4) >= kBigitSize) {
```

File: double-conversion/bignum.cc [hunk 9]
```diff
@@ -201,24 +181,28 @@ void Bignum::AddBignum(const Bignum& other) {
...
 
-  EnsureCapacity(1 + (std::max)(BigitLength(), other.BigitLength()) - exponent_);
+  EnsureCapacity(1 + (std::max)(BigitLength(), other.BigitLength()) - exponent_.get());
   Chunk carry = 0;
-  int bigit_pos = other.exponent_ - exponent_;
+  int bigit_pos = other.exponent_.get() - exponent_.get();
   DOUBLE_CONVERSION_ASSERT(bigit_pos >= 0);
-  for (int i = 0; i < other.used_digits_; ++i) {
-    Chunk sum = RawBigit(bigit_pos) + other.RawBigit(i) + carry;
+  for (int i = used_bigits_; i < bigit_pos; ++i ) {
```
Comment (floitsch) [inline double-conversion/bignum.cc]:
```suggestion
  for (int i = used_bigits_; i < bigit_pos; ++i) {
```

File: double-conversion/utils.h [hunk 10]
```diff
@@ -221,6 +219,11 @@ class Vector {
 
   T& last() { return start_[length_ - 1]; }
 
+  void pop_back() {
+    DOUBLE_CONVERSION_ASSERT( !is_empty() );
```
Comment (floitsch) [inline double-conversion/utils.h]:
```suggestion
    DOUBLE_CONVERSION_ASSERT(!is_empty());
```

**Comment by ColinH** [comment]
So I removed the `ExponentWrapper`, and accepted all of your suggestions (didn't even know that GitHub feature yet), but now there's the GoogleBot; do you know what exactly the problem is? Did I forget something?

**Comment by floitsch** [comment]
No idea where the cla-issue comes from, but I have verified the Authors, and I don't see any problem.

**Comment by floitsch** [comment]
thanks!

**Pull request merged**